lib/spell-check-manager.coffee

```diff
@@ -0,0 +1,384 @@
+    multirange = require 'multi-integer-range'
+
+    # For every registered spellchecker, we need to find out the ranges in the
+    # text that the checker confirms are correct or indicates is a misspelling.
+    # We keep these as separate lists since the different checkers may indicate
+    # the same range for either and we need to be able to remove confirmed words
+    # from the misspelled ones.
+    correct = new multirange.MultiRange([])
+    incorrects = []
+
+    for checker in @checkers
+      # We only care if this plugin contributes to checking spelling.
+      if not checker.isEnabled() or not checker.providesSpelling(args)
+        continue
+
+      # Get the results which includes positive (correct) and negative (incorrect)
+      # ranges. If we have an incorrect range but no correct, everything not
+      # in incorrect is considered correct.
+      results = checker.check(args, text)
+
+      if results.invertIncorrectAsCorrect and results.incorrect
+        # We need to add the opposite of the incorrect as correct elements in
+        # the list. We do this by creating a subtraction.
+        invertedCorrect = new multirange.MultiRange([[0, text.length]])
+        removeRange = new multirange.MultiRange([])
+        for range in results.incorrect
+          removeRange.appendRange(range.start, range.end)
+        invertedCorrect.subtract(removeRange)
+
+        # Everything in `invertedCorrect` is correct, so add it directly to
+        # the list.
+        correct.append invertedCorrect
+      else if results.correct
+        for range in results.correct
+          correct.appendRange(range.start, range.end)
+
+      if results.incorrect
+        newIncorrect = new multirange.MultiRange([])
+        incorrects.push(newIncorrect)
+
+        for range in results.incorrect
+          newIncorrect.appendRange(range.start, range.end)
+
+    # If we don't have any incorrect spellings, then there is nothing to worry
+    # about, so just return and stop processing.
+    misspellings = []
+
+    if incorrects.length is 0
+      return {id: args.id, misspellings}
+
+    # Build up an intersection of all the incorrect ranges. We only treat a word
+    # as being incorrect if *every* checker that provides negative values treats
+    # it as incorrect. We know there are at least one item in this list, so pull
+    # that out. If that is the only one, we don't have to do any additional work,
+    # otherwise we compare every other one against it, removing any elements
+    # that aren't an intersection which (hopefully) will produce a smaller list
+    # with each iteration.
+    intersection = null
+    index = 1
+
+    for incorrect in incorrects
+      if intersection is null
+        intersection = incorrect
+      else
+        intersection.append(incorrect)
+
+    # If we have no intersection, then nothing to report as a problem.
+    if intersection.length is 0
+      return {id: args.id, misspellings}
```

Comment:
Same as https://github.com/atom/spell-check/pull/120/files#r74591382, can we return `misspellings: []` here?